File: 2023/aoc23.py
import sys

days = {
    1: __import__('01'),
    2: __import__('02'),
#   3: __import__('03'),
#   4: __import__('04'),
#   5: __import__('05'),
#   6: __import__('06'),
#   7: __import__('07'),
#   8: __import__('08'),
#   9: __import__('09'),
#   10: __import__('10'),
#   11: __import__('11'),
#   12: __import__('12'),
#   13: __import__('13'),
#   14: __import__('14'),
#   15: __import__('15'),
#   16: __import__('16'),
#   17: __import__('17'),
#   18: __import__('18'),
#   19: __import__('19'),
#   20: __import__('20'),
#   21: __import__('21'),
#   22: __import__('22'),
#   23: __import__('23'),
#   24: __import__('24'),
#   25: __import__('25'),
}

def solve(day, file=sys.stdin):
    puzzle_input = day.get_input(file)
    print(day.part1(puzzle_input))
    print(day.part2(puzzle_input))

while True:
    if len(sys.argv) > 1:
        d = int(sys.argv[1])
        solve(days[d])
        break
    try:
        d = int(input('Day? '))
        with open(f'{d:02}/input.txt') as f:
            solve(days[d], file=f)
    except KeyboardInterrupt:
        break
    except Exception as e:
        sys.exit(f'Error: {e}')

File: 2023/aoc23.exs
defmodule Main do
  def days(1), do: Day01
  def days(2), do: Day02
  def days(3), do: Day03
  def days(4), do: Day04
  def days(5), do: Day05
  def days(6), do: Day06
  def days(7), do: Day07
  def days(8), do: Day08
  def days(9), do: Day09
  def days(10), do: Day10
  def days(11), do: Day11
  def days(12), do: Day12
  def days(13), do: Day13
  def days(14), do: Day14
  def days(15), do: Day15
  def days(16), do: Day16
  def days(17), do: Day17
  def days(18), do: Day18
  def days(19), do: Day19
  def days(20), do: Day20
  def days(21), do: Day21
  def days(22), do: Day22
  def days(23), do: Day23
  def days(24), do: Day24
  def days(25), do: Day25

  def solve(day, file) do
    input = day.get_input(file)
    input |> day.part1() |> IO.inspect()
    input |> day.part2() |> IO.inspect()
  end

  def loop() do
    d =
      case IO.gets("> ") |> String.trim() do
        "q" -> System.halt(0)
        "" -> loop()
        n -> n
      end
    {:ok, file} = File.open("#{String.pad_leading(d, 2, "0")}/input.txt")
    d |> String.to_integer() |> days() |> solve(file)
    File.close(file)
    loop()
  end
end

case List.first(System.argv) do
  nil -> Main.loop()
  d -> d |> String.to_integer() |> Main.days() |> Main.solve(:stdio)
end
